signer/local_cosigner.go

```diff
@@ -69,92 +31,162 @@ type LocalCosignerConfig struct {
+	total               uint8
+	address             string
+	// peers       map[int]CosignerPeer
+	localsigner *LocalSoftSignThresholdEd25519Signature // TODO: Change the mutex so that you can you can pass ThresholdEd25519Signature
 }
 
 func NewLocalCosigner(cfg LocalCosignerConfig) *LocalCosigner {
+	// TODO: localsigner should be passed as a parameter in the cfg rather than constructed here. We could add a method in the new LocalSoftSignThresholdEd25519Signature file to return this config such as func NewLocalThresholdEd25519Signature(key CosignerKey, rsaKey PrivateKey, total, threshold int) LocalSoftSignThresholdEd25519Signature { which could then be called and added to the LocalCosignerConfig from cmd/horcrux/cmd/cosigner.go
+
+	localsigner := LocalSoftSignThresholdEd25519Signature{
+		Key:       cfg.CosignerKey,
+		RsaKey:    cfg.RsaKey,
+		HrsMeta:   make(map[HRSTKey]HrsMetadata),
+		Peers:     make(map[int]CosignerPeer),
+		Total:     cfg.Total,
+		Threshold: cfg.Threshold,
+	}
+
+	LastSignStateStruct := LastSignStateStruct{
+		LastSignStateMutex: sync.Mutex{},
+		LastSignState:      cfg.SignState,
+	}
 	cosigner := &LocalCosigner{
-		key:           cfg.CosignerKey,
-		lastSignState: cfg.SignState,
-		rsaKey:        cfg.RsaKey,
-		hrsMeta:       make(map[HRSTKey]HrsMetadata),
-		peers:         make(map[int]CosignerPeer),
-		total:         cfg.Total,
-		threshold:     cfg.Threshold,
-		address:       cfg.Address,
+		//key:         cfg.CosignerKey,
+		//pubKeyBytes: []byte{},
+		LastSignStateStruct: &LastSignStateStruct,
+		total:               cfg.Total,
+		address:             cfg.Address,
+		localsigner:         &localsigner,
 	}
 
+	// TODO: Delete this print statements:
+	// fmt.Println("\n", "LocalCosigner")
+	// fmt.Printf("%+v\n", cosigner)
+
+	// fmt.Println("\n", "Local Signer")
+	// fmt.Printf("%+v\n", cosigner.localsigner)
+
 	for _, peer := range cfg.Peers {
-		cosigner.peers[peer.ID] = peer
+		cosigner.localsigner.Peers[peer.ID] = peer
 	}
 
 	// cache the public key bytes for signing operations
-	switch ed25519Key := cosigner.key.PubKey.(type) {
+	switch ed25519Key := cosigner.localsigner.Key.PubKey.(type) {
 	case tmCryptoEd25519.PubKey:
-		cosigner.pubKeyBytes = make([]byte, len(ed25519Key))
-		copy(cosigner.pubKeyBytes, ed25519Key[:])
+		cosigner.localsigner.PubKeyBytes = make([]byte, len(ed25519Key))
+		copy(cosigner.localsigner.PubKeyBytes, ed25519Key[:])
+		cosigner.pubKeyBytes = cosigner.localsigner.PubKeyBytes
 	default:
 		panic("Not an ed25519 public key")
 	}
 
 	return cosigner
 }
 
+func (cosigner *LocalCosigner) SaveLastSignedState(signState SignStateConsensus) error {
+	return cosigner.LastSignStateStruct.LastSignState.Save(signState, &cosigner.LastSignStateStruct.LastSignStateMutex)
+}
+
 // GetID returns the id of the cosigner
-// Implements Cosigner interface
+// Implements cosigner interface
 func (cosigner *LocalCosigner) GetID() int {
-	return cosigner.key.ID
+	return cosigner.localsigner.Key.ID
```

Comment:
For this, we can either populate the ID as an int param to the `LocalCosigner` struct, or make an interface method `GetID() int` in the `ThresholdEd25519Signature`, and the specific threshold signature implementation can return it.

Reply:
Went for GetID